pkg/gardenlet/controller/managedseed/add.go

```diff
@@ -0,0 +1,327 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedseed
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/gardener/gardener/charts"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
+	"github.com/gardener/gardener/pkg/controllerutils/mapper"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	confighelper "github.com/gardener/gardener/pkg/gardenlet/apis/config/helper"
+	"github.com/gardener/gardener/pkg/utils"
+	contextutil "github.com/gardener/gardener/pkg/utils/context"
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+const (
+	// ControllerName is the name of this controller.
+	ControllerName = "managedseed"
+
+	// GardenletDefaultKubeconfigSecretName is the default name for the field in the Gardenlet component configuration
+	// .gardenClientConnection.KubeconfigSecret.Name
+	GardenletDefaultKubeconfigSecretName = "gardenlet-kubeconfig"
+	// GardenletDefaultKubeconfigBootstrapSecretName is the default name for the field in the Gardenlet component configuration
+	// .gardenClientConnection.BootstrapKubeconfig.Name
+	GardenletDefaultKubeconfigBootstrapSecretName = "gardenlet-kubeconfig-bootstrap"
+)
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(
+	mgr manager.Manager,
+	cfg config.GardenletConfiguration,
+	gardenCluster cluster.Cluster,
+	seedCluster cluster.Cluster,
+	shootClientMap clientmap.ClientMap,
+	imageVector imagevector.ImageVector,
```

Comment:
```suggestion
	gardenCluster cluster.Cluster,
	seedCluster cluster.Cluster,
```

The other fields should be set directly when constructing the `Reconciler` struct.

Reply:
These fields are needed by Actuator `	r.Actuator = newActuator(
		gardenCluster.GetConfig(),
		gardenCluster.GetAPIReader(),
		gardenCluster.GetClient(),
		seedCluster.GetClient(),
		shootClientMap,
		NewValuesHelper(&cfg, imageVector),
		gardenCluster.GetEventRecorderFor(ControllerName+"-controller"),
		r.ChartsPath,
		r.GardenNamespaceShoot,
	)`

Reply:
Fine, but you can still access these fields from the `Reconciler` struct instead of passing them here

Reply:
With this change `ChartsPath` and `GardenNamespaceShoot ` can also be removed from `Reconciler` struct as it is only used by `Actuator`. We can directly set pass these values to Actuator while constructing the `Reconciler` struct. WDYT?

Reply:
The actuator is only some internal struct in the reconciler and we expose it only for unit testability. Hence, I would not like to start constructing this in the general `controller/add.go` file. It should be constructed in the `AddToManager` of the `Reconciler` if it is not already set.
We also have other places where the `Reconciler` has fields which it does not use in the `Reconcile` function, but I don't think this is a problem/bad thing.